src/objc/AudioPreprocessing.m

```diff
@@ -0,0 +1,365 @@
+                                   code:MLModelErrorGeneric
+                               userInfo:@{NSLocalizedDescriptionKey: @(*errorMsg)}];
+    }
+}
+
+static inline float hzToMel(float freq) {
+    return 1127.0 * log(1 + freq/700.0f);
+}
+
+
+@interface PowerSpectrum : NSObject
+
+@property (readonly) float* PowerSpectrum;
+
+-(id)initWithFFTLength:(NSUInteger)fftLength withFrameLength:(NSUInteger)frameLength;
+-(void)applyHammingWindow:(float*)signal numSamples:(NSUInteger)numSamples;
+-(void)calculatePowerSpectrum:(float*)signal numSamples:(NSUInteger)numSamples;
+
+@end
+
+
+@implementation PowerSpectrum {
+    float *_hammingWindow, *_workingFrame;
+
+    DSPSplitComplex _spectrum;
+    FFTSetup _fftSetup;
+    
+    NSUInteger _frameLength, _fftLength;
+}
+
+
+-(id)initWithFFTLength:(NSUInteger)fftLength withFrameLength:(NSUInteger)frameLength {
+    
+    self = [super init];
+    
+    _fftLength = fftLength;
+    _frameLength = frameLength;
+    
+    _hammingWindow = (float*)calloc(frameLength, sizeof(float));
+    [self _generateHammingWindow:frameLength];
+    
+    _fftSetup = vDSP_create_fftsetup(log2(fftLength*2), FFT_RADIX2);
+    _spectrum.realp = (float*)calloc(fftLength, sizeof(float));
+    _spectrum.imagp = (float*)calloc(fftLength, sizeof(float));
+    _PowerSpectrum = (float*)calloc(fftLength/2 + 1, sizeof(float));
+    _workingFrame = (float*)calloc(fftLength, sizeof(float));
+    
+    return self;
+}
+
+-(void)dealloc {
+    free(_hammingWindow);
+    vDSP_destroy_fftsetup(_fftSetup);
+    free(_spectrum.realp);
+    free(_spectrum.imagp);
+    free(_PowerSpectrum);
+    free(_workingFrame);
+}
+
+-(void)applyHammingWindow:(float*)signal numSamples:(NSUInteger)numSamples {
+    NSAssert(numSamples == _frameLength, @"Num samples and frame length must be the same.");
+    vDSP_vmul(_hammingWindow, 1, signal, 1, signal, 1, numSamples);
+}
+
+-(void)calculatePowerSpectrum:(float*)signal numSamples:(NSUInteger)numSamples {
+    // Copy signal into working frame.
+    memcpy(_workingFrame, signal, numSamples * sizeof(float));
+
+    // Pad signal with 0 to fftLength.
+    memset(&_workingFrame[numSamples], 0, (_fftLength - numSamples) * sizeof(float));
+
+    // Produce packed input necessary for vDSP_fft to work.
+    vDSP_ctoz((DSPComplex *)_workingFrame, 2, &_spectrum, 1, _fftLength/2);
+
+    // Perform the forward FFT.
+    vDSP_fft_zrip(_fftSetup, &_spectrum, 1, log2(_fftLength), kFFTDirection_Forward);
+
+    float scale = 0.5; // Factor of 2 needed due to vDSP_fft_zrip implementation detail.
+    vDSP_zvabs(&_spectrum, 1, _PowerSpectrum, 1, _fftLength/2);
+    vDSP_vsmul(_PowerSpectrum, 1, &scale, _PowerSpectrum, 1, _fftLength/2);
+}
+
+-(void)_generateHammingWindow:(NSUInteger)frameLength {
+    // Don't use vDSP_hamm_window because it uses a slightly different implementation
+    for(size_t i = 0; i < frameLength; i++) {
+        _hammingWindow[i] = 0.5 - 0.5*cos((2*M_PI*i)/(frameLength));
```

Comment:
Another place where you may as well use `float` values (and `cosf`) to minimize work given that you're only storing `float` precision anyway

Reply:
Couldn't find a float version of `M_PI`, but I have changed the two `0.5` to `0.5f` and the `cos` to `cosf`.